nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+		}
+	}
+	e.Status = stResolved
+	e.Flags = flags
+	e.putTime = now
+	e.validUntil = now.Add(time.Second * time.Duration(minTtl))
+	e.reply = *reply
+
+	if shouldNotify {
+		close(e.waitChan) // notify all the waiters by closing the channel
+	}
+}
+
+// wait until a valid reply is set in the cache
+func (e *cacheEntry) waitReply(request *dns.Msg, timeout time.Duration, now time.Time) (*dns.Msg, error) {
+	if e.Status == stResolved {
+		return e.getReply(request, now)
+	}
+
+	if timeout > 0 {
+		select {
+		case <-e.waitChan:
+			return e.getReply(request, now)
+		case <-time.After(time.Second * timeout):
+			return nil, errTimeout
+		}
+	}
+
+	return nil, errCouldNotResolve
+}
+
+func (e *cacheEntry) close() {
+	if e.Status == stPending {
+		close(e.waitChan)
+	}
+}
+
+// entriesSlice is used for sorting entries
+type cacheEntriesSlice []*cacheEntry
+
+func (p cacheEntriesSlice) Len() int           { return len(p) }
+func (p cacheEntriesSlice) Less(i, j int) bool { return p[i].validUntil.Before(p[j].validUntil) }
+func (p cacheEntriesSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+type cacheKey struct {
+	question dns.Question
+	protocol dnsProtocol
+}
+
+type entries map[cacheKey]*cacheEntry
+
+// Cache is a thread-safe fixed capacity LRU cache.
+type Cache struct {
+	Capacity int
+
+	entries entries
+	lock    sync.RWMutex
+}
+
+// NewCache creates a cache of the given capacity
+func NewCache(capacity int) (*Cache, error) {
+	if capacity <= 0 {
+		return nil, errInvalidCapacity
+	}
+	c := &Cache{
+		Capacity: capacity,
+		entries:  make(entries, capacity),
+	}
+	return c, nil
+}
+
+// Clear removes all the entries in the cache
+func (c *Cache) Clear() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entries = make(entries, c.Capacity)
+}
+
+// Purge removes the old elements in the cache
+func (c *Cache) Purge(now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.removeOldest(c.Capacity, now)
+}
+
+// Add adds a reply to the cache.
+func (c *Cache) Put(request *dns.Msg, reply *dns.Msg, proto dnsProtocol, flags uint8, now time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	question := request.Question[0]
+	key := cacheKey{question, proto}
+	if ent, found := c.entries[key]; found {
+		ent.setReply(reply, flags, now)
+	} else {
+		// If we will add a new item and the capacity has been exceeded, make some room...
+		if len(c.entries) >= c.Capacity {
+			c.removeOldest(1, now)
+		}
+		c.entries[key] = newCacheEntry(&question, reply, stResolved, now)
+	}
+}
+
+// Look up for a question's reply from the cache.
+// If no reply is stored in the cache, it returns a `nil` reply and no error. The caller can then `Wait()`
+// for another goroutine `Put`ing a reply in the cache.
+func (c *Cache) Get(request *dns.Msg, proto dnsProtocol, now time.Time) (reply *dns.Msg, err error) {
```

Comment:
>> what happens if we send requests to ns1.google.com announcing, for example, a buffer size of 65535 bytes?

> WeaveDNS doesn't do that; why would it?

As you said, it is 4K, and it could be ok to do so when no cache was used, but caching that response would be useless for some clients. A client could not accept such a long response, and no cache entry would be useful. Yes, you could truncate a long response in the cache, but I think the point in having a cache is not only to avoid requests to upstream servers but to avoid this kind of computations...

By the way, current WeaveDNS implementation does not truncate responses to clients if they support less than 4K, right? I think the recursive resolver always uses 4Kb, and I don't remember to have seen the part of the code that truncates responses...

> In sum: the current algorithm is
> 
>  * For a UDP request:
>    * Do we have a cached UDP response? If so send that, truncating if necessary
>       * If not, ask upstream, cache the result, and send it downstream, truncating if necessary
>    * For a TCP request:
>     * Do we have a cached TCP response? If so send that
>       * If not, ask upstream over TCP, cache the result and send it downstream

No, my proposal never truncates responses. As I said, I think it is not a good thing to do... The only things I modify in responses is the TTL (and, in some cases, some shuffling)

> * For a request:
>    * Do we have a cached response? If so, send that, truncating if necessary
>      * If not, ask upstream*, cache the result, and send it downstream, truncating if necessary

I think these algorithms could be classified more like:

1) to have full responses in the cache, obtained either with TCP or with UDP, and truncate as needed, according to client needs.
2) to have the short responses, obtained with UDP, some (hopefully, just a few) with the TC bit set and some (hopefully, just a few) full responses (obtained with TCP), and never process the packet.

So your algorithm pros are (and correct me if I'm wrong)

+ it would save one RTT to the upstream server for long responses >512bytes.
+ it would save a byte per cache entry as well as the entries for short, truncated responses when long responses are also present.

but the cons would be

- it would be useless if the upstream server does not support EDNS
- hopefully, it will only make a difference for a tiny fraction of queries (as most queries fit in the original 512-bytes payload limit), and when previous condition is met. 
- it would require implementing the truncation heuristics (eg, you have to decide which RR should be removed...)
- it could lead to many truncations (for example, if the response in the cache for `google.com NS` is 4Kb and all WeaveDNS clients are not-EDNS capable or with short buffers) 

In conclusion, I think your algorithm is nice but it adds more complexity and just saves us some bytes in some corner cases... :|